dial.go

```diff
@@ -308,6 +342,105 @@ func (d FdDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
 	return conn, nil
 }
 
+// Dial makes FdDialer satisfy the Dialer interface.
+func (d FdDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	d.RequiredProtocolInfo.Auth = NoAuth
+
+	dialer := ProtocolDialer{
+		Dialer: fdDialer{
+			fd: d.Fd,
+		},
+		RequiredProtocolInfo: d.RequiredProtocolInfo,
+	}
+
+	return dialer.Dial(ctx, opts)
+}
+
+type AuthDialer struct {
+	// Dialer is a base dialer.
+	Dialer Dialer
+	// Authentication options.
+	Auth     Auth
+	Username string
+	Password string
+}
+
+// Dial makes AuthDialer satisfy the Dialer interface.
+func (d AuthDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	conn, err := d.Dialer.Dial(ctx, opts)
+	if err != nil {
+		return conn, err
+	}
+	greeting := conn.Greeting()
+	if greeting.Salt == "" {
+		conn.Close()
+		return nil, fmt.Errorf("failed to authenticate: " +
+			"an invalid connection without salt")
+	}
+
+	if d.Username == "" {
+		return conn, nil
+	}
+
+	protocolAuth := conn.ProtocolInfo().Auth
+	if d.Auth == AutoAuth {
+		if protocolAuth != AutoAuth {
+			d.Auth = protocolAuth
+		} else {
+			d.Auth = ChapSha1Auth
+		}
+	}
+
+	if err := authenticate(conn, d.Auth, d.Username, d.Password,
+		conn.Greeting().Salt); err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("failed to authenticate: %w", err)
+	}
+	return conn, nil
+}
+
+type ProtocolDialer struct {
+	// Dialer is a base dialer.
+	Dialer Dialer
+	// RequiredProtocol contains minimal protocol version and
+	// list of protocol features that should be supported by
+	// Tarantool server. By default, there are no restrictions.
+	RequiredProtocolInfo ProtocolInfo
+}
+
+// Dial makes ProtocolDialer satisfy the Dialer interface.
+func (d ProtocolDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	conn, err := d.Dialer.Dial(ctx, opts)
+	if err != nil {
+		return conn, err
+	}
+
+	protocolConn := protocolConn{
+		Conn:         conn,
+		protocolInfo: d.RequiredProtocolInfo,
+	}
+
+	protocolConn.protocolInfo, err = identify(protocolConn)
+	if err != nil {
+		protocolConn.Close()
+		return nil, fmt.Errorf("failed to identify: %w", err)
+	}
+
+	err = checkProtocolInfo(d.RequiredProtocolInfo, protocolConn.protocolInfo)
+	if err != nil {
+		protocolConn.Close()
+		return nil, fmt.Errorf("invalid server protocol: %w", err)
+	}
+
+	if d.RequiredProtocolInfo.Auth != AutoAuth {
+		protocolConn.protocolInfo.Auth = d.RequiredProtocolInfo.Auth
+	} else if protocolConn.protocolInfo.Auth == AutoAuth {
+		protocolConn.protocolInfo.Auth = ChapSha1Auth
+	}
```

Comment:
It does not have sense in fact, we need just to pass the value from the server.
```suggestion
```

Reply:
The main problem is:

1) If `Auth` passed to the `OpenSslDialer` it should overwrite the response from the server.
2) If `Auth` (for `OpenSslDialer`) is `AutoAuth`, then the response is used. If the response auth is also `AutoAuth`, then `ChapSha1Auth` is used.

So here I tried to solve the first problem. At the start of the `OpenSslDialer`-s `Dial` function, if `Auth` is specified, it stored in the `RequiredProtocolInfo`. Then it should overwrite the server response.
Is this a bad solution for all dialers? If the user explicitly specified the `Auth` method in the `RequiredProtocolInfo`,  should we use it regardless of the server response?

https://github.com/tarantool/go-tarantool/blob/6ba01ff28ee01dcc1f22f9fa508110cdb10e66a0/dial.go#L239-L246

Reply:
We pass the authentication type here:
https://github.com/tarantool/go-tarantool/blob/8969661b74e89215c87d0739f4db53d374d5eb47/dial.go#L260

It will be used here:
https://github.com/tarantool/go-tarantool/blob/8969661b74e89215c87d0739f4db53d374d5eb47/dial.go#L366-L368

The default would be `ChapSha1Auth`, I don't see a problem here.

Reply:
But `ProtocolInfo.Auth` then will contain the value, that differs from the real `Auth`, used by the connection. Is it ok?

Reply:
Used by which connection? ProtocolInfo.Auth contains a value that server prefers.

Reply:
So value in the `ProtocolInfo.Auth` does not matter, if we specified the `Auth` field in the `OpenSslDialer`? There is no point in checking `ProtocolInfo.Auth` in tests?